aomi/model/__init__.py

```diff
@@ -0,0 +1,626 @@
+        """Copies a secret into a particular location"""
+        for sfile in self.secrets():
+            src_file = hard_path(sfile, self.opt.secrets)
+            if not os.path.exists(src_file):
+                raise aomi.exceptions.IceFile("%s secret not found at %s" %
+                                              (self, src_file))
+
+            dest_file = "%s/%s" % (tmp_dir, sfile)
+            dest_dir = os.path.dirname(dest_file)
+            if not os.path.isdir(dest_dir):
+                os.mkdir(dest_dir, 0o700)
+
+            shutil.copy(src_file, dest_file)
+            log("Froze %s %s" % (self, sfile), self.opt)
+
+    def resources(self):
+        """List of included resources"""
+        return [self]
+
+    def grok_state(self, obj):
+        """Determine the desired state of this
+        resource based on data present"""
+        if 'state' in obj:
+            my_state = obj['state'].lower()
+            if my_state != 'absent' and my_state != 'present':
+                raise aomi \
+                    .exceptions \
+                    .Validation('state must be either "absent" or "present"')
+
+        self.present = obj.get('state', 'present').lower() == 'present'
+
+    def validate(self, obj):
+        """Base validation method. Will inspect class attributes
+        to dermine just what should be present"""
+        if 'tags' in obj and not isinstance(obj['tags'], list):
+            raise aomi.exceptions.Validation('tags must be a list')
+
+        if self.present:
+            check_obj(self.required_fields, self.name(), obj)
+
+    def name(self):
+        """A Friendly Name for our Resource"""
+        return self.__doc__.split('\n')[0]
+
+    def __str__(self):
+        return "%s %s" % (self.name(), self.path)
+
+    def obj(self):
+        """Returns the Python dict/JSON object representation
+        of this Secret as it is to be written to Vault"""
+        return self._obj
+
+    # note that this is going to be implemented by subclasses
+    def secrets(self):  # pylint: disable=no-self-use
+        """Returns a list of secrets which may be used used
+        locally by this Vault resource"""
+        return []
+
+    def __init__(self, obj, opt):
+        self.grok_state(obj)
+        self.validate(obj)
+        self.path = None
+        self.existing = None
+        self._obj = {}
+        self.tags = obj.get('tags', [])
+        self.opt = opt
+
+    def fetch(self, vault_client):
+        """Populate internal representation of remote
+        Vault resource contents"""
+        result = self.read(vault_client)
+        if result:
+            if isinstance(result, dict) and 'data' in result:
+                self.existing = result['data']
+            else:
+                self.existing = result
+        else:
+            self.existing = None
+
+    def sync(self, vault_client):
+        """Update remove Vault resource contents if needed"""
+        if self.present and not self.existing:
```

Comment:
I took a look into this and while it would result in a few lines of code for the `log` output, in order to have this kind of logic for full CRUD operations it would end up too complex. I created https://github.com/Autodesk/aomi/issues/121 to handle this in the future.